[3.3.5] Scripts/Stratholme: Baroness Anastari rework

hiMalleus opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Added possess ability with base on sniffs

**Target branch(es):** 3.3.5/master

- [X] 3.3.5

**Issues addressed:** Closes #20335


**Tests performed:** (Does it build, tested in-game, etc.)
- Builds
- Tested in game with a party of 2


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Update zone_grizzly_hills.cpp

Commit:
Added spells to Enraged Spirits in SMV

Commit:
Update zone_shadowmoon_valley.cpp

Commit:
Indentation issues fixed

Commit:
Indentation issues

Commit:
More Indentation changes

Commit:
Many more indentation changes

Commit:
tabs2spaces and whitespaces

Commit:
Update victim before events update

Commit:
broggok death poison cloud

Added poison cloud on his death

Commit:
whitespaces

Commit:
Revert "broggok death poison cloud"

This reverts commit 7a9cf546ebcd3e2234e56ae36985dbde6df5a43b.

Commit:
Revert "whitespaces"

This reverts commit 3dc191c69f3a40be301a8dbcbad7af3836a64ad3.

Commit:
Winterfin quest and wrong space

changed his state from Aggressive to Defensive as shown on video https://www.youtube.com/watch?v=apkZvnlkLX8 on minute 1:57, His cage door now opens for 20 seconds calculated from video and fixed a misstyped space on Broggok script

Commit:
Escape from the Winterfin Caverns

Fixed not getting credit for escorting the npc until the end.

Commit:
Enraged Spirit timers and missing spell

Commit:
baroness anastari rework

Reworked her script in order to make Possess ability work.

Commit:
Revert "Enraged Spirit timers and missing spell"

This reverts commit 950bfef865023e7f91150d83fcb28d6f27b5d76d.

Commit:
Revert "Escape from the Winterfin Caverns"

This reverts commit 9035095b1ddcee2628447fbf3b7ffe383d3c2639.

Commit:
Revert "Winterfin quest and wrong space"

This reverts commit 148605948b5602f2c3dcc3457c5fd8508daacd30.

Commit:
Revert "Revert "whitespaces""

This reverts commit e804f2ba83e13c4d28e1497fa4851097ae0cfc6c.

Commit:
Revert "Revert "broggok death poison cloud""

This reverts commit 7e3483a30e88d211666d6cce89aab50c66701a7b.

Commit:
Revert "whitespaces"

This reverts commit 3dc191c69f3a40be301a8dbcbad7af3836a64ad3.

Commit:
Revert "broggok death poison cloud"

This reverts commit 7a9cf546ebcd3e2234e56ae36985dbde6df5a43b.

Commit:
Revert "Update victim before events update"

This reverts commit f72a27f40a743000a7eda11b48128935ecd1feb0.

Commit:
Revert "tabs2spaces and whitespaces"

This reverts commit 58185022cf7ca1932b85153b984098e4c6b41843.

Commit:
Revert "Many more indentation changes"

This reverts commit 4c429edf9da585ab3b7625a484377bd1d117d335.

Commit:
Revert "More Indentation changes"

This reverts commit eed7dd1fd1030516e084201df6f58b43b2a163b0.

Commit:
Revert "Indentation issues"

This reverts commit d10df9c70287c6948eb51f173271d596f63a76c6.

Commit:
Revert "Indentation issues fixed"

This reverts commit 0e30de853b582923dbbccde333f298b8c8759fd5.

Commit:
Revert "Update zone_shadowmoon_valley.cpp"

This reverts commit 524623ded0ede23beaa8eb878de0e8f98496e64a.

Commit:
Revert "Added spells to Enraged Spirits in SMV"

This reverts commit 4d6118af8ee40725be9ae6da81af4b29b6e4b0a3.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 9fee43ea89ead373e7f0d7cf1e1644ced8f73065.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit cf518840386e4640553a5f4f8a869cf83c9e326e.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 8f667778a757c3717ecf10c6b429cb189d69175f.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 68e813a7b0e7983892d62c2b2479335ca701b397.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 3ea077ade1f966d010c4ac4e544a2ee1b5dea525.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 1d2630674fcfc7d132b9c18452c5bada73868596.

Commit:
Revert "Update zone_grizzly_hills.cpp"

This reverts commit 09e3bc46549daa1f541e10e958a31dbbf8c8cfc5.

Ovahlord commented:
You should convert all the timers to eventMap or eventScheduler since timers are outdated.

hiMalleus commented:
I've seen several Vanilla bosses scripted with timers instead of eventMap or eventScheduler, what is the advantage of events over timers really?

Ovahlord commented:
coding standards. Timers are not used anymore. The scripts where you have seen timers are probably outdated and in need of a rewrite / refactor. Take a look at newer scripts instead like in icecrown citadel or ulduar.

hiMalleus commented:
Ok thanks for insight. I'll change things when i get home.

Commit:
Baroness Anastari Events

Changed from deprecated timers to events

Commit:
a whitespace

Commit:
check fix

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -32,50 +32,59 @@ enum Spells
-        //uint32 Possess_Timer;
+        
+        bool invisible;
+        Unit* possessedTarget;
```
Never store WorldObject and derived classes pointers in scripts

hiMalleus replied:
why not? I was going to use objectguid but i don't know the diference of having a global pointer...

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -32,50 +32,59 @@ enum Spells
-        uint32 Silence_Timer;
-        //uint32 Possess_Timer;
+        
+        bool invisible;
```
Always initialize all class member fields in the constructor

hiMalleus replied:
will do

Warpten replied:
@jackpoz do static analysis pick up static initializers correctly now?

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +97,73 @@ class boss_baroness_anastari : public CreatureScript
+                    if (!possessedTarget->HasAura(SPELL_POSSESSED))
+                        _events.ScheduleEvent(EVENT_INVISIBLE, Seconds(0));
+
+                    if (possessedTarget->GetHealthPct() <= 50)
```
is this supposed to be a "else if" to avoid scheduling the same event twice ?

hiMalleus replied:
You're right, I will test with OR to see if it also works.

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +97,73 @@ class boss_baroness_anastari : public CreatureScript
+                        break;
+                    case EVENT_SPELL_POSSESS:
+                        possessedTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false);
+                        if (possessedTarget->IsAlive())
```
missing NULL check on possessedTarget (by taking a look at this code it seems it could be NULL)

hiMalleus replied:
It will never be NULL because when you select target random it has an offset of 1. If no target is selected then possessedtarget will never be alive

Ovahlord replied:
just use if (possessedTarget && possessedTarget->IsAlive()) to make sure that you wont get a nullpointer here to avoid crashes.

Warpten replied:
This will crash if people go in alone. So yes.

hiMalleus replied:
I tested alone and it didn't crash but ok I will change it

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -32,50 +32,59 @@ enum Spells
+                {
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                    possessedTarget->SetObjectScale(1.0f);
```
what happens if the target is dead with object scale set to 1.5f ?

Kittnz replied:
wouldn't it just scale back? or am i missing something

hiMalleus replied:
It does scale back. The reset is only to make sure when the party wipes if someone is still possessed then dispells him

Ovahlord replied:
I think the scaling itself is handled by a spell instead. This also overrides auras that increase/change the scale of a unit.

hiMalleus replied:
Nothing on sniffs show scaling spells

Commit:
Changes

Set private and protected variables, created ObjectGUID to avoid storing  WorldObject and derived classes pointers in scripts and fixed a few validations

Keader commented:
Already has a work in progress that will fix this boss: #17860

Keader commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,74 @@ class boss_baroness_anastari : public CreatureScript
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
+                Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false);    // Random target to be possessed
```
it should be inside of EVENT_SPELL_POSSESS

tkrokli replied:
I presume that will be `case EVENT_SPELL_POSSESS:`
But only the line above, or more related code?

offl replied:
Only the line above. Same for the comment below.

tkrokli replied:
Very well. I can move line 101 into `case EVENT_SPELL_POSSESS:`.
I still need to know where line 102 goes before I move it.

offl replied:
```cpp
case EVENT_INVISIBLE:
```

Keader commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,74 @@ class boss_baroness_anastari : public CreatureScript
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
+                Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false);    // Random target to be possessed
+                Unit* possessedTarget = me->GetMap()->GetPlayer(_possessedTargetGuid);          // When there's a possessed target
```
It should be inside of some event...

Keader commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,74 @@ class boss_baroness_anastari : public CreatureScript
+                Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false);    // Random target to be possessed
+                Unit* possessedTarget = me->GetMap()->GetPlayer(_possessedTargetGuid);          // When there's a possessed target
+
+                if (_invisible)
```
you should transform it in a event, if you need check each 1 sec or each X time.
Current local is wrong.

Keader commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,74 @@ class boss_baroness_anastari : public CreatureScript
+                        _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, Seconds(4));
+                        break;
+                    case EVENT_SPELL_BANSHEECURSE:
+                        DoCast(me->GetVictim(), SPELL_BANSHEECURSE);
```
DoCastVictim(SPELL_BANSHEECURSE);

Keader commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,74 @@ class boss_baroness_anastari : public CreatureScript
+                        break;
+                    case EVENT_SPELL_BANSHEECURSE:
+                        DoCast(me->GetVictim(), SPELL_BANSHEECURSE);
+                        _events.ScheduleEvent(EVENT_SPELL_BANSHEECURSE, Seconds(18));
```
_events.Repeat(Seconds(18));

Faq commented:
@hiMalleus can you correct requested changes?

tkrokli commented:
if @hiMalleus does not respond, TC members are able to edit these files, but we do prefer that the original author makes the changes (mostly because of the learning perspective).

offl commented:
> TC members are able to edit these files

Seems the time has come.

tkrokli commented:
OK, I can add Keader's last 2 suggestions
```cpp
DoCastVictim(SPELL_BANSHEECURSE);
```
and
```cpp
_events.Repeat(Seconds(18));
```
but I don't know the solution to his first 3 comments. Would you like to suggest what goes into those 3 before I add those 2 changes?

offl commented:
@Keader any tips?

tkrokli commented:
I guess it will take a while for him to pick up on this, so I'll just add the 2 known changes. Not a big deal. We'll just do it in 2 commits instead of 1. :-)

Commit:
Update boss_baroness_anastari.cpp

- add 2 of the 5 changes suggested by Keader

Commit:
Update boss_baroness_anastari.cpp

- move line 101 according to Keader's suggestion

tkrokli commented:
Hmm. The downside to moving line 101 and adding a blank is that it caused comment https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r144872379 and comment https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r144872563 to be hidden in this thread. They need to be followed up with 2 new changes.

tkrokli commented:
OK, so I broke the switch statement with my latest change. Need to fix that too. Wrong formatting, perhaps?

Commit:
Revert the bad commit

offl commented:
```cpp
                    case EVENT_SPELL_POSSESS:
                        if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))    // Random target to be possessed
                        {
                            if (possessTarget && possessTarget->IsAlive())
                            {
                                me->CastStop();
                                DoCast(possessTarget, SPELL_POSSESS);
                                DoCast(possessTarget, SPELL_POSSESSED);
                                possessTarget->SetObjectScale(1.5f);
                                me->SetReactState(REACT_PASSIVE);
                                me->SetVisible(false);
                                _invisible = true;
                                _possessedTargetGuid = possessTarget->GetGUID();
                            }
                            else
                                _events.ScheduleEvent(EVENT_SPELL_POSSESS, Seconds(20), Seconds(30));
                        }
                        break;
```
```cpp
                    case EVENT_INVISIBLE:
                        if (Unit* possessedTarget = me->GetMap()->GetPlayer(_possessedTargetGuid))          // When there's a possessed target
                        {
                            possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
                            possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
                            possessedTarget->SetObjectScale(1.0f);
                            me->SetReactState(REACT_AGGRESSIVE);
                            me->SetVisible(true);
                            _invisible = false;
                            _events.ScheduleEvent(EVENT_SPELL_POSSESS, Seconds(20), Seconds(30));
                        }
                        break;
```
Try this. You must remove lines 101-102 of course.

Commit:
Update boss_baroness_anastari.cpp

- apply requested changes

Commit:
Update boss_baroness_anastari.cpp

- remove lines 101-102 (+ empty line) also

tkrokli commented:
Hmm, maybe we could add
```cpp
Unit* possessedTarget = me->GetMap()->GetPlayer(_possessedTargetGuid);
```
in `void UpdateAI(uint32 diff) override` / after `_events.Update(diff);` ?

tkrokli commented:
Oh. My bad, I think we need to move the code below `if (_invisible)` to the `case EVENT_INVISIBLE:` section.

offl commented:
No, https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r144872379 should be resolved now.

tkrokli commented:
Sure, but I think this is about https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r144872563 instead and the lines below it, which causes the travis-ci error: `use of undeclared identifier 'possessedTarget' `

tkrokli commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +94,77 @@ class boss_baroness_anastari : public CreatureScript
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
+                if (_invisible)
+                    if (!possessedTarget->HasAura(SPELL_POSSESSED) || possessedTarget->GetHealthPct() <= 50)
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:102:26:
 fatal error: use of undeclared identifier 'possessedTarget'
                    if (!possessedTarget->HasAura(SPELL_POSSESSED) || possessedTarget->GetHealthPct() <= 50)
                         ^
1 error generated.
```

offl commented:
In enum
```cpp
    EVENT_CHECK_POSSESSED = 6,
```
After line 136
```cpp
                    case EVENT_CHECK_POSSESSED:
                        if (_invisible)
                        {
                            if (Unit* possessedTarget = me->GetMap()->GetPlayer(_possessedTargetGuid))
                            {
                                if (!possessedTarget->HasAura(SPELL_POSSESSED) || possessedTarget->GetHealthPct() <= 50)
                                    _events.ScheduleEvent(EVENT_INVISIBLE, Seconds(0));
                                else
                                    _events.ScheduleEvent(EVENT_CHECK_POSSESSED, Seconds(1));
                            }
                        }
                        break;
```
After line 131
```cpp
                                _events.ScheduleEvent(EVENT_CHECK_POSSESSED, Seconds(0));
```
Remove lines 101-104
I'm not quite familiar with EventMap. And I think things like this must be done by devs, to avoid flood :c

Commit:
add extra event to follow up required changes

offl commented:
@jackpoz could you please approve these changes?

jackpoz commented:
was https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r139643691 addressed ? it doesn't look like that to me

tkrokli commented:
Looks like hiMalleus went Missing in Action after comment https://github.com/TrinityCore/TrinityCore/pull/20336#discussion_r139679209. We will need to address the request `Always initialize all class member fields in the constructor`.

tkrokli commented:
The user hiMalleus has been deleted recently.

jackpoz commented on src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp:
```diff
@@ -88,37 +95,86 @@ class boss_baroness_anastari : public CreatureScript
+                switch (eventId)
+                {
+                    case EVENT_SPELL_BANSHEEWAIL:
+                        DoCast(me->GetVictim(),SPELL_BANSHEEWAIL);
```
I see DoCastVictim() is used too when casting a spell to a victim, maybe it makes sense to use the same chosen syntax all over the code when the victim is the desired target

jackpoz commented:
The author of this PR disappeared. Anyone with C++ knowledge would like to take over this PR ? Otherwise we'll close it in a few weeks

jackpoz commented:
I added new labels with "Closed-*" prefix to easily find closed PRs that weren't merged in case someone wants to reopen them (no need to open a new PR in that case)

Aokromes commented:
Anyone can takeover this pr?

tkrokli commented:
Unless a real coder volunteers, I could open a new PR from my account just to be able to perform requested changes.

Then again, it would be more beneficial for someone with more time on their hands and a real interest in coding / scripting to start off with a script in a single file like this (possibly "[good first issue]" ?)

jackpoz commented:
Ask here to reopen the PR if anyone wants to work on it
